hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/conf/ConfigurationGenerator.java

```diff
@@ -88,7 +88,8 @@ public Configuration generate(SourceOfRandomness random, GenerationStatus genera
                 e.printStackTrace();
             }
         }
-        //ConfigTracker.freshMap();  // --> Comment out if we want to incrementally collect exercised config set.
+        //ConfigTracker.clearConfigMap();  // --> Comment out if we want to incrementally collect exercised config set.
+        ConfigTracker.clearSetConfigMap();  // Necessary for replay !
```

Comment:
So we only need to clean the setMap instead of the configMap?
Can you explain why we need to do so?

Reply:
If we don't do so, in the next round a config previously in setmap might be only `get` but not `set` by the test case, causing the diff in replay file.

Reply:
Okay, this makes sense. Thanks for the explanation.

Reply:
@hwang-pku Would it be better to record the setMap (the same way as we record the configMap) of the failed execution instead of doing the clear here?

I'm afraid that even if we clear the set here, there could still have differences between the replay and fuzz round.

Reply:
why would there still be differences between replay and fuzz?

Reply:
> If we don't do so, in the next round a config previously in setmap might be only `get` but not `set` by the test case, causing the diff in replay file.

Sorry I still have confusion.

The replay goal will only generate value for those parameters in `configMap`, so as long as our recorded `configMap` is correct, then we must generate the same parameter keys. 
Even a config parameter previously is by `get` not by `set`, this parameter will be put in the `configMap` in that round, and if that round fails, the recorded `configMap` will have this config; if later this config is `set`, the `configMap` will remove it and if this round fails, the recorded `configMap` will not have this config.

So why `setMap` is still related here?

Reply:
For example, in the previous round `foo.bar` is set, so it is not in `configMap`. Thus, for this round we are not going to generate anything for this config. There's no problem regarding this. 
However, consider if in the current round `foo.bar` is only get but not set, if we do not clear `setConfigMap`, the `currentConfContext` dumped by the `ConfuzzGuidance` will not include `foo.bar`. However, `foo.bar` would appear in replay since we use a fresh `setConfigMap` in replay, causing an inconsistency.

Reply:
if in the current round foo.bar is only get but not set:
1. Fuzz goal will NOT generate `foo.bar` because it is in the `setMap`
2. Then `currentConfContext` will also not have this `foo.bar`
3. Then replay only reads and generates parameters in `replayConfContext`: https://github.com/xlab-uiuc/confuzz/blob/master/confuzz-guidance/src/main/java/edu/illinois/confuzz/internal/ReplayData.java#L33 and replay has nothing to do with `setConfigMap`. 

Why the issue comes from `setConfigMap` instead of the inconsistency of `replayConfContext` and `currentConfContext`?

Reply:
No, but the `ConfContext` for Replay will have `foo.bar` since although it is not generated,  it is collected by ConfigTracker.configMap, and is involved in the comparison afterwards.

Reply:
Here `ConfContext` you mean `replayConfContext` or `currentConfContext`?

Reply:
`replayConfContext` contains all parameters that replay round need to generate
`currentConfContext` contains all parameters and values that failure round in fuzzing.

So replay goal will generate all parameters based on `replayConfContext` and compare them with `currentConfContext`.
So if there is a difference, which means that `replayConfContext.keySet()` will also diff with `currentConfContext.keySet()`?

Reply:
What replay goal does is:
1. Generate config based on `replayConfContext`
2. Run the test
3. Collect a `confContext`
4. compare this `confContext` with `currentConfContext`
What I meant is that the `confContext` collected in step 3 would contain `foo.bar`, which is different from `currentConfContext`.

Reply:
Oh, I see. This makes more sense to me. Thanks a lot for the explanation.